src/v/s3/client.cc

```diff
@@ -527,14 +532,18 @@ ss::future<> client_pool::stop() {
 ///         are in use)
 ss::future<client_pool::client_lease> client_pool::acquire() {
     gate_guard guard(_gate);
-    if (_pool.empty()) {
+    while (_pool.empty() && !_gate.is_closed()) {
         if (_policy == client_pool_overdraft_policy::wait_if_empty) {
-            co_await _cvar.wait([this] { return !_pool.empty(); });
+            co_await _cvar.wait();
```

Comment:
it looks like this might be a case where a semaphore with an initial value of 0 might fix some issues. in particular it gives you semaphore like functionality, but doesn't suffer from "dropped signals" if no waiters are present. I don't really know if it makes sense here, but just a thought.

Reply:
I'm still using `wait` that accepts predicate in `stop` method (it doesn't have this problem in that case). Semaphore will make the code a bit more verbose.